### PR TITLE
Compute `BboxLoss` weights and pose keypoint strides on matched anchors only

### PR DESCRIPTION
**Pull request by JESUSROYETH**
Both the bbox-loss weight and the pose keypoint-stride division work on all ~8400 anchors and then keep only the foreground-matched ones. Doing the masking first gives the same result over a much smaller tensor.

**Bbox-loss weight.** `BboxLoss`, `RotatedBboxLoss` and `v8OBBLoss` sum `target_scores` over the class dim for every anchor and then index the foreground rows:

```python
weight = target_scores.sum(-1)[fg_mask].unsqueeze(-1)
```

Masking first sums only the matched rows, which is the same weight:

```python
weight = target_scores[fg_mask].sum(-1, keepdim=True)
```

(the OBB angle-loss site in `v8OBBLoss` uses `target_scores[fg_mask].sum(-1)` without the unsqueeze, matching its existing shape.)

**Pose keypoint stride.** `v8PoseLoss` and `PoseLoss26` divide the whole `(BS, N_anchors, 17, 2)` selected-keypoints tensor by stride, but only `selected_keypoints[masks]` is ever read:

```python
selected_keypoints[..., :2] /= stride_tensor.view(1, -1, 1, 1)
...
if masks.any():
    gt_kpt = selected_keypoints[masks]
```

So the division moves into the masked branch, dividing only the matched anchors and skipping it entirely when there are none:

```python
if masks.any():
    gt_kpt = selected_keypoints[masks]
    gt_kpt[..., :2] /= stride_tensor.view(1, -1).expand(masks.shape[0], -1)[masks][:, None, None]
```

The adjacent `target_bboxes /= stride_tensor` has the same full-then-mask shape but stays as it is on purpose: it is a caller-owned tensor divided in place, and much smaller `(BS, N_anchors, 4)`, while `selected_keypoints` is a local temporary read only through `[masks]`, so moving its division is the safe and worthwhile half.

**Deleted:** the class-dim sum over all anchors in the three bbox-weight sites, and the full-tensor keypoint/stride division in both pose losses (skipped now when no anchor is matched).

**Validation** (ultralytics 8.4.103, RTX 4060):

- Bit-identical. I ran the real `BboxLoss.forward` and `calculate_keypoints_loss` on fixed inputs at realistic foreground sparsity (~1.5%), and the losses match before and after with max abs diff `0` for both. A 1-epoch train of yolo11n-pose, yolo26n-pose, yolo11n-obb and yolo11n detect runs clean.
- The two reductions are ~6-7x faster in isolation on GPU (the bbox-weight sum ~6.9x, the keypoint/stride division ~6.3x). This runs on every training step for the detection and pose families, but the absolute saving is small next to a full step, so it is a clean, identical-output cleanup of wasted work rather than a big end-to-end training speedup.

No new test: the loss runs on every training step and the existing pose/OBB/detect training tests already cover these paths; the output does not change.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Computes bounding-box and pose keypoint loss weights and strides only for matched anchors, improving alignment between loss calculations and foreground assignments.

### 📊 Key Changes
- Updates standard and rotated `BboxLoss` weighting to select foreground anchors before summing target scores.
- Applies pose keypoint stride normalization after selecting matched anchors in both pose loss implementations.
- Updates angle-loss weighting to use target scores from matched anchors only.

### 🎯 Purpose & Impact
- Avoids unnecessary computation on unmatched anchors and keeps loss weights consistent with foreground targets.
- Improves stride handling for pose keypoints across varying anchor matches.
- Affects detection, rotated bounding-box, pose, and angle-loss training behavior without changing public APIs.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JESUSROYETH, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms that your changes have been received, and an engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The code changes are clean. The masked bbox-weight calculations preserve the original shapes and values, and the pose stride normalization correctly selects the stride corresponding to each matched anchor before applying it to the gathered keypoints.

**Comment by glenn-jocher** [comment]
Thanks for the focused optimization—the mask-first reductions appear equivalent and CI is green. Please add a small regression test covering a multi-image batch with sparse matches across different strides, plus the no-foreground pose path, so the new masked indexing is explicitly protected.

**Comment by glenn-jocher** [comment]
Verified bitwise equivalence independently before merge: all five changed sites (bbox/dfl weights, OBB angle weight, both pose stride divisions) produce max abs deviation 0.0 vs main at unit level (fp32/fp16, both DFL branches, realistic ~1.5% fg sparsity) and end-to-end on yolo11n detect, yolo11n-pose, and yolo26n-pose including the RLE branch — losses bitwise identical. So this is a clean optimization, not a bugfix; I've relabeled `enhancement` accordingly. Net −4 lines in loss.py, no guards, and grep confirms the five sites are ALL occurrences of the pattern (segment inherits BboxLoss.forward). CPU microbench shows ~1.6x on the weight computation; the honest 'small vs a training step' framing in the body matches what I measured.

**Comment by glenn-jocher** [comment]
Slimmed to the minimal change: dropped the test additions — bitwise equivalence is verified and recorded above. Diff is now net −4.
